egressgateway: fix manager logic

jibi opened the pull request:
This PR refactors the egress gateway manager in order to provide a
single `reconcile()` method, responsible for adding and removing entries
to and from the egress_policy map, that will be invoked on all events
received by the manager.

In addition to this, the manager will now wait for the k8s cache to be
fully synced before running its first reconciliation, in order to always
have the egress_policy map in a consistent state with the k8s
configuration.

Fixes: #17380
Fixes: #17753

pchaigno commented on cilium/cmd/bpf_egress_get.go:
```diff
@@ -43,13 +42,11 @@ var bpfEgressGetCmd = &cobra.Command{
 		}
 
-		fmt.Println(value.String())
+		fmt.Printf("%s %s", value.GetGatewayIP(), value.GetEgressIP())
```
Nit: Could we put this into a `EgressPolicyVal4.String()` method as before?

pchaigno commented on cilium/cmd/bpf_egress_list.go:
```diff
@@ -15,20 +18,45 @@ import (
+	egressListUsage = "List egress policy entries.\n" + lpmWarningMessage
 )
 
+type EgressPolicy struct {
```
```suggestion
type egressPolicy struct {
```

pchaigno commented on cilium/cmd/bpf_egress_list.go:
```diff
@@ -15,20 +18,45 @@ import (
-		if err := egressmap.EgressMap.Dump(bpfEgressList); err != nil {
-			Fatalf("error dumping contents of map: %s\n", err)
+		if err := egressmap.OpenEgressMaps(); err != nil {
+			if errors.Is(err, fs.ErrNotExist) {
```
This calls into `ebpf.OpenMap` (cilium/cilium) which relies on `ebpf.LoadPinnedMap` (cilium/ebpf). But as far as I could see `ebpf.LoadPinnedMap` never returns `ErrNotExist`. Did I miss something?

jibi replied:
`ebpf.LoadPinnedMap` will call into `sys.ObjGet`, which in turn will call `sys.BPF`, and the BPF syscall there can return `ENOENT`:

```
vagrant@k8s1:~$ sudo cilium bpf egress list
Cannot find egress gateway bpf maps
```

pchaigno replied:
Ah, of course! :facepalm: Thanks!

pchaigno commented on cilium/cmd/bpf_egress_list.go:
```diff
@@ -41,13 +69,23 @@ var bpfEgressListCmd = &cobra.Command{
 			fmt.Fprintf(os.Stderr, "No entries found.\n%v\n", lpmWarningMessage)
 		} else {
-			TablePrinter("SRC IP & DST CIDR", "EGRESS INFO", bpfEgressList)
+			printEgressList(bpfEgressList)
```
I'm guessing you're not using `TablePrinter` here because it shows map contents separated in only two columns and you'd prefer four columns? Could we extend `TablePrinter` instead to support multiple columns? (Probably best done in another PR.)

jibi replied:
~Whops, just a leftover from another branch (but in this case we can definitely use the `TablePrinter`, will switch to it :+1: )~

edit: sorry, I got confused, let's keep it like this for now, I'll look into extending the `TablePrinter` helper to support multiple columns in a separate PR

pchaigno commented on pkg/maps/egressmap/egress.go:
```diff
@@ -4,79 +4,72 @@
-	DestCIDR types.IPv4
-}
+const (
+	PolicyMapName = "cilium_egress_policy_v4"
```
Won't the map name change cause a short connectivity blip on upgrade for egress gateway connections? Probably fine given it's a beta feature, but should we document it?

jibi replied:
> Won't the map name change cause a short connectivity blip on upgrade for egress gateway connections? Probably fine given it's a beta feature

yep, I remember asking for that and we said it was acceptable given the beta status

> but should we document it?

will add a note in the docs :+1:

pchaigno commented on pkg/egressgateway/manager.go:
```diff
@@ -191,104 +152,93 @@ func (manager *Manager) OnDeleteEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
+
+nextPolicyKey:
+	for policyKey, policyVal := range egressPolicies {
+		for _, policyConfig := range manager.policyConfigs {
+			for _, endpoint := range manager.epDataStore {
+				if !policyConfig.selectsEndpoint(endpoint) {
+					continue
+				}
```
Nit: These 5 lines could be moved outside the `egressPolicies` loop, to avoid computing the same thing for each egress policy.

jibi replied:
I don't think it's possible as for each entry in the bpf map we necessarily need to test all k8s policies. The pseudocode would be:

```
nextEntry:
for each entry in the egress_policy map {
    for each policy in k8s egressnatpolicies {
        if policy matches entry {
            // we found one k8s policy that matches the current BPF entry, move to the next one
            continue nextEntry
        }
    }

    // the current BPF entry is not backed by any k8s policy, delete it
    egressmap.RemoveEgressPolicy()
}
```

(while if we start looping from k8s policies, we would not be able to tell if a given BPF entry needs to be deleted or not)

I'll add a comment to clarify this

pchaigno commented on pkg/egressgateway/manager.go:
```diff
@@ -191,104 +152,93 @@ func (manager *Manager) OnDeleteEndpoint(endpoint *k8sTypes.CiliumEndpoint) {
-	return nil
+
+	manager.addMissingEgressRules()
+	manager.removeUnusedEgressRules()
```
The order of those two calls seems to matter for updated entries, as we would otherwise remove them before updating them. Is it worth a comment?

jibi replied:
Yep, order matters here, will add a comment :+1:

pchaigno commented on pkg/egressgateway/manager_privileged_test.go:
```diff
@@ -0,0 +1,262 @@
+		{ep2IP, destCIDR, egressIP2, egressIP2},
+	})
+
+	// Update the endpoint labels for polikcy-1 in order for it to not be a match
```
Nit:
```suggestion
	// Update the endpoint labels for policy-1 in order for it to not be a match
```

pchaigno commented on bpf/lib/egress_gateway.h:
```diff
@@ -1,12 +1,22 @@
 /* Copyright (C) 2021 Authors of Cilium */
 
-#ifndef __LIB_EGRESS_POLICIES_H_
-#define __LIB_EGRESS_POLICIES_H_
```
Can we keep this and the filename as is? It was done on purpose since this same file will inherit the SRv6 egress policies as well.

pchaigno reviewed (commented):
Mostly nits.

jibi committed:
docs: add a note on egress gateway upgrade impact for 1.11

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

jibi committed:
bpf: rename egress policy map and its fields

to make it more clear it's related to the egress gateway policies

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

kkourt commented on Documentation/cmdref/cilium_bpf_egress.md:
```diff
@@ -23,6 +23,6 @@ Manage the egress routing rules
 * [cilium bpf egress delete](../cilium_bpf_egress_delete)	 - Delete egress entries
 * [cilium bpf egress get](../cilium_bpf_egress_get)	 - Get egress entries
-* [cilium bpf egress list](../cilium_bpf_egress_list)	 - List egress entries
+* [cilium bpf egress list](../cilium_bpf_egress_list)	 - List egress policy entries
```
[Elsewhere](https://github.com/cilium/cilium/pull/17868#discussion_r749211292), @pchaigno mentioned that there are SRv6 egress policies coming. Will these policies be on the same map or another one? In the latter case, maybe we should refer to this map as egress gw policies?

jibi replied:
(not sure, but this seems to me a bit out of scope for the current PR)

kkourt replied:
I thought that since we are chancing the name anyway, we could use "List egress gateway policy entries". Follow up is fine as well.

kkourt commented on cilium/cmd/bpf_egress_list.go:
```diff
@@ -41,13 +69,23 @@ var bpfEgressListCmd = &cobra.Command{
+func printEgressList(egressList []egressPolicy) {
+	w := tabwriter.NewWriter(os.Stdout, 5, 0, 3, ' ', 0)
+
+	fmt.Fprintln(w, "Source IP\tDestination CIDR\tEgress IP\tGateway IP")
```
Thanks! a description of the various fields was definitely missing!

kkourt commented on pkg/egressgateway/manager.go:
```diff
@@ -34,94 +20,96 @@ var (
+// sync with k8s and then runs the first reconciliation.
+func (manager *Manager) runReconciliationAfterK8sSync() {
+	go func() {
+		for range time.NewTicker(time.Second * 1).C {
```
Do we need to grab the lock before checking if the cache is synced?

I'm not familiar with the details of the code, and I'm not clear on what the `mutex` actually protects. There are two potential reasons for needing the lock here:
1. we are using an internal field of `manager` (k8sCacheSyncedChecker). Does this needs to be protected by the mutex?
1. again, without knowing too much about the code, it seems possible that the cache will become unsynced after we break and before we take the lock. Do we care about that? If yes, can we use the lock to ensure that we always run reconcile with a synced cache?

jibi replied:
A bit of context here: the `k8sCacheSyncedChecker` field is just an interface for an object that implements the `K8sCacheIsSynced` method. In this case we are passing the `Daemon` to the egress gw manager, which implements it as:

https://github.com/cilium/cilium/blob/b12ecd16302da5943379f1d7e088e07a61c2f3e4/daemon/cmd/daemon.go#L1037-L1046

This method just checks if the `k8sCachesSynced` channel is closed, which happens only after the initial k8s sync (i.e. after the agent has received all the k8s state it cares about from the apiserver during initialization). After that it will remain closed (we cannot go back to an unsycned state). 

Because of this, I don't think we need to acquire the manager's lock (which is supposed to protect the manager internal state from concurrent accesses) to run this check

kkourt reviewed (approved):
Some non-blocking comments from my side.

joamaki commented on pkg/egressgateway/manager.go:
```diff
@@ -34,94 +20,96 @@ var (
+// sync with k8s and then runs the first reconciliation.
+func (manager *Manager) runReconciliationAfterK8sSync() {
+	go func() {
+		for range time.NewTicker(time.Second * 1).C {
```
You'll need to `Stop()` the ticker, otherwise (I think) this leaks resources.You could just do loop with time.Sleep instead...

Though in this case, since `K8sCacheIsSynced` is just waiting on `d.k8sCachesSynced`, 
why not just do a blocking receive on `d.k8sCachesSynced` instead?

E.g. add something like:
```
func (d *Daemon) WaitUntilK8sCacheIsSynced() {
  _, _ = <-d.k8sCachesSynced
}
```

jibi replied:
Good shout, switched to `WaitUntilK8sCacheIsSynced`  :+1:

joamaki commented on pkg/maps/egressmap/egress.go:
```diff
@@ -4,79 +4,72 @@
+		return fmt.Errorf("cannot apply egress policy: %w", err)
+	}
+
+	logger.Info("Applied egress policy")
```
Hm, wonder if it'd be better to log both the error and success case in the same place? When I looked at this function I immediately wondered if we log the returned error somewhere (yep, in OnAddEgressPolicy) and if we log it with enough details.

jibi replied:
Yep, makes sense. I moved all the logging into the manager and also got rid of `egressmap.{ApplyEgressPolicy, RemoveEgressPolicyMap}` since those methods where just wrapping a single function call (`egressmap.EgressPolicyMap.{Update, Delete}`)

joamaki commented on pkg/maps/egressmap/egress.go:
```diff
@@ -4,79 +4,72 @@
+	log.WithFields(logrus.Fields{
+		logfields.SourceIP:        sourceIP,
+		logfields.DestinationCIDR: destCIDR,
+	}).Info("Removed egress policy")
```
Same issue here with logging as in ApplyEgressPolicy.

jibi committed:
daemon: add WaitUntilK8sCacheIsSynced method

which will block the caller until the agent has fully sync its
k8s cache.

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

qmonnet reviewed (approved):
I only did had a quick look at the changes, and all look in order to me. I'll look more in-depth if I can find a moment, but I trust the other people who looked in more details.

As for the few changes falling under `docs-structure`, this looks all good.

jibi committed:
egressgateway: refactor manager logic

This commit refactors the egress gateway manager in order to provide a
single `reconcile()` method  which will be invoked on all events
received by the manager. This method is responsible for adding and
removing entries to and from the egress policy map.

In addition to this, the manager will now wait for the k8s cache to be
fully synced before running its first reconciliation, in order to always
have the egress_policy map in a consistent state with the k8s
configuration.

Fixes: #17380
Fixes: #17753
Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

jibi committed:
maps: switch egressmap to cilium/ebpf package

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

pchaigno commented on bpf/lib/egress_policies.h:
```diff
@@ -34,25 +44,16 @@ is_cluster_destination(struct iphdr *ip4, __u32 dst_id, __u32 tunnel_endpoint)
 }
 
-# define lookup_ip4_egress_endpoint(sip, dip) \
-	egress_lookup4(&EGRESS_MAP, sip, dip)
```
FYI, I'll have to revert all changes to this file in the SRv6 PR, except for the map/key name changes.

EDIT: Actually, I can preserve this function name change as well, doesn't make a different for SRv6 for now.
